keras_cv/layers/preprocessing/channel_shuffle.py

```diff
@@ -52,15 +51,10 @@ def __init__(self, groups=3, seed=None, **kwargs):
         self.groups = groups
         self.seed = seed
 
-    def _channel_shuffling(self, images):
-        unbatched = images.shape.rank == 3
-
-        if unbatched:
-            images = tf.expand_dims(images, axis=0)
-
-        height = tf.shape(images)[1]
-        width = tf.shape(images)[2]
-        num_channels = images.shape[3]
+    def augment_image(self, image, transformation=None):
+        height = tf.shape(image)[0]
+        width = tf.shape(image)[1]
+        num_channels = image.shape[2]
```

Comment:
you will need to use tf.shape(image)[3].  We can't use `.shape` in graph mode.

Reply:
@LukeWood This gives the following error:

```
Using a symbolic `tf.Tensor` as a Python `bool` is not allowed in Graph execution. Use Eager execution or decorate this function with @tf.function.
```

Using `tf.cond` seems like the only option. 
Also,  `.shape` seems to be working fine as part of a model too.

Reply:
The first bug makes sense - you do need to use tf.cond there.  as for the .shape usage - there are cases where it will cause issues.  I'm fairly sure we can't use .shape inside of call methods - @qlzh727 can you confirm this?

Reply:
You can actually use .shape here. Check the fchollet comment here. https://github.com/keras-team/keras-cv/pull/124#discussion_r815002419

Reply:
This would solve everything actually.

Reply:
@LukeWood This seems to solve all the other errors as well. I believe we can ensure that the channels dimension will always be populated. Do you think this can be merged as is?

Of course, I'll reduce the width and height to one single call, however.

Reply:
@qlzh727 @LukeWood (cc. @quantumalaviya )
I checked channel shuffle on actual training for augmenting the input image. It works without any issue. [HERE](https://www.kaggle.com/code/ipythonx/tf-hybrid-efficientnet-swin-transformer-gradcam). So, as a preprocessing layer or augmenting layer, it's completely safe with the current implementation. No code changes may be required. 

Now, if you followed a discussion over [HERE](https://github.com/keras-team/keras-cv/pull/124#issuecomment-1042876194), it later extended to work with intermediate layers (i.e `shuffle-net`). But it wasn't checked earlier by passing it as the intermediate layer. I'm checking it now. As the `num_channels` is known for input value and may not be known for intermediate layer output. So, it may cause issues. 

So, I tried to run the following model but it raised another issue for `tf.random_shuffle`. 

```python
input_shape = (28, 28, 3)
model = keras.Sequential(
    [
        keras.Input(shape=input_shape),
        ChannelShuffle(groups=3), # OK 
        layers.Conv2D(32, kernel_size=(3, 3), activation="relu"),
        ChannelShuffle(groups=4), # Not OK
        layers.MaxPooling2D(pool_size=(2, 2)),
        ChannelShuffle(groups=8), # Not OK
        layers.Conv2D(64, kernel_size=(3, 3), activation="relu"),
        layers.MaxPooling2D(pool_size=(2, 2)),
        layers.Flatten(),
        layers.Dropout(0.5),
        layers.Dense(num_classes, activation="softmax"),
    ]
)
````
```
StagingError: in user code:
....
 LookupError: gradient registry has no entry for RandomShuffle
```

---

- https://github.com/tensorflow/tensorflow/issues/6269#issuecomment-465850464.
- https://github.com/tensorflow/tensorflow/issues/22505
- https://stackoverflow.com/q/70875986/9215780 

What are the main causes of it?

Reply:
Awesome - thank you for testing it in training.  Great to know!